File: 747-min-cost-climbing-stairs/min-cost-climbing-stairs.js
/**
 * @param {number[]} cost
 * @return {number}
 */
var minCostClimbingStairs = function(cost) {
    const n = cost.length

    if (n === 0) return 0
    if (n === 1) return cost[0]
    if (n === 2) return Math.min(cost[0], cost[1])

    const dp = []

    for (let i = 0; i < n; i++) {
        if (i === 0 || i === 1) {
            dp[i] = cost[i]
            continue;
        }
        dp[i] = cost[i] + Math.min(dp[i-1], dp[i-2])
    }

    return Math.min(dp[n-1], dp[n-2])
};

File: 747-min-cost-climbing-stairs/min-cost-climbing-stairs.py
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        def min_cost(i):
            if i <= 1:
                return 0

            if i in memo:
                return memo[i]

            down_one = cost[i-1] + min_cost(i-1)
            down_two = cost[i-2] + min_cost(i-2)
            memo[i] = min(down_one, down_two)
            return memo[i]

        memo = {}
        return min_cost(len(cost))